### DISCUSSION TITLE
Customizing UI

### DISCUSSION THREAD

**Discussion by josephconradt** [excalidraw/excalidraw#3854] [Q&A] [answered]
Is it possible to override Excalidraw UI elements without modifying source code? My goal is to use Material UI for the buttons/dialogs/controls etc. UI appears to be done in the `LayerUI` component. Any documentation or advice?

BTW this is a fantastic project. And I'm hoping to contribute when I have a better understanding of the codebase.

Thanks in advance.

**Comment by ad1992** [comment]
Hi @josephconradt 👋🏻 currently we don't support customizing the UI elements out of the box. In the future, we might decouple all the UI elements and create a library out of it where we might be able to support customization. But as per your use case, since you want to use material UI, for now, you will have to fork the repo and update all the UI elements!
Is it possible to achieve the same look and feel just by overriding CSS ? Then would suggest to do that instead of forking the repo and updating the code.

**Reply by josephconradt** [reply]
Thank you for the quick answer! Decoupling the UI would be a cool enhancement. Potentially exposing actions (for example) "select draw tool" that could be called from a custom component - any button could invoke it. Allowing more manipulation for advanced use cases. I'll probably end up forking for now. Thanks again.

**Reply by josephconradt** [reply]
I'm going to look into the `ActionManager`... perhaps I can invoke actions from a completely custom UI? 🤔

**Reply by ad1992** [reply]
yeah thats something we have in the pipeline https://github.com/excalidraw/excalidraw/issues/3007

**Comment by josephconradt** [comment, marked as answer]
For anyone curious. I was able to find a solution. You can change the app state directly from the API. Try inspecting the app state object as it contains many options from the lefthand panel.

Set the current tool to pencil

```js
ExcalidrawAPI.updateScene({
  appState: {
    elementType: "freedraw"
  }
});
```

**Reply by claws99** [reply]
Helllo, I'm a beginner. Can you please elaborate on how to do this? Does this require forking the repo?

**Comment by mikepianka** [comment]
I am looking to do something similar to what @josephconradt was interested in. I've been poking around and have not found a great solution to do the following things. I would like to achieve it just through CSS if possible to not have to create a separate fork. Any suggestions appreciated, thanks in advance!

### Hide the Library and Help icons
I ended up with these CSS selectors that are working ok. I don't love the help selector. Is there a better way to hide UI elements that aren't exposed through props or is this still the preferred way?
```css
/* library button */
label.ToolIcon.ToolIcon_type_floating.ToolIcon__library.ToolIcon_size_medium {
  display: none;
}

/* help button */
div.layer-ui__wrapper__footer-right.zen-mode-transition > button {
  display: none;
}
```

### Add custom floating action buttons that call Excalidraw actions
I'd like to add a couple floating action buttons for common tasks, like the Save as Image dialog. I looked through the `appState` object and didn't notice any properties to trigger opening the Save as Image dialog from within it. Any advice on triggering arbitrary Excalidraw actions?

### Customize non-primary colors
I was able to use the `--color-primary` CSS variable to change the highlight color in the drawing toolbar. Are the blue and red colors seen in other parts of the app, like the `Save as Image` dialog, also exposed in a similar customizable way?
![image](https://user-images.githubusercontent.com/47948499/160479817-aa5025ab-cfef-4d01-a2c0-54410b8d0455.png)
